render-ws-java-client/src/main/java/org/janelia/render/client/MultiSemFlatFieldCorrectionClient.java

```diff
@@ -0,0 +1,195 @@
+		final ClientRunner clientRunner = new ClientRunner(args) {
+			@Override
+			public void runClient(final String[] args) throws IOException {
+
+				final Parameters parameters = new Parameters();
+				parameters.parse(args);
+				LOG.info("runClient: entry, parameters={}", parameters);
+
+				final MultiSemFlatFieldCorrectionClient client = new MultiSemFlatFieldCorrectionClient(parameters);
+				client.correctTiles();
+			}
+		};
+		clientRunner.run();
+	}
+
+	public MultiSemFlatFieldCorrectionClient(final Parameters parameters) {
+		this.params = parameters;
+		this.renderClient = new RenderDataClient(parameters.multiProject.baseDataUrl, parameters.multiProject.owner, parameters.multiProject.project);
+	}
+
+	public void correctTiles() throws IOException {
+		final List<StackId> stacks = params.multiProject.stackIdWithZ.getStackIdList(renderClient);
+
+		// TODO: iterate by z-layer first to re-use the flat field estimate for all stacks?
+		for (final StackId stack : stacks) {
+			final DoubleSummaryStatistics zStatistics = getZStatistics(stack);
+			final StackMetaData stackMetaData = renderClient.getStackMetaData(stack.getStack());
+			renderClient.setupDerivedStack(stackMetaData, stack.getStack() + params.targetStackSuffix);
+
+			for (int z = (int) zStatistics.getMin(); z <= (int) zStatistics.getMax(); z++) {
+				final ResolvedTileSpecCollection tileSpecs = renderClient.getResolvedTiles(stack.getStack(), (double) z);
+
+				for (final TileSpec tileSpec : tileSpecs.getTileSpecs()) {
+					final ImageProcessor ip = loadImageTile(tileSpec);
+					final int sfov = extractSfovNumber(tileSpec);
+					final ImageProcessor flatFieldEstimate = loadFlatFieldEstimate(Math.min(z, params.flatFieldConstantFromZ), sfov);
+
+					applyFlatFieldCorrection(ip, flatFieldEstimate);
+
+					patchTileSpec(tileSpec);
+					saveImage(ip, tileSpec);
+				}
+
+				renderClient.saveResolvedTiles(tileSpecs, stack.getStack() + params.targetStackSuffix, (double) z);
+			}
+			renderClient.setStackState(stack.getStack() + params.targetStackSuffix, StackMetaData.StackState.COMPLETE);
+		}
+	}
+
+	private DoubleSummaryStatistics getZStatistics(final StackId stack) throws IOException {
+		final List<Double> zValues = renderClient.getStackZValues(stack.getStack());
+		final DoubleSummaryStatistics zStatistics = zValues.stream().mapToDouble(Double::doubleValue).summaryStatistics();
+		LOG.info("Considering {} with {} z-layers", stack, zStatistics.getCount());
+		return zStatistics;
+	}
+
+	private ImageProcessor loadFlatFieldEstimate(final int z, final int sfov) {
+		final Path imagePath = Path.of(params.flatFieldLocation, String.format(params.flatFieldFormat, z, sfov));
+		final String imageUrl = "file:" + imagePath;
+		final ImageLoader.LoaderType loaderType = ImageLoader.LoaderType.IMAGEJ_DEFAULT;
+		return CACHE.get(imageUrl, 0, false, false, loaderType, null);
+	}
+
+	private int extractSfovNumber(final TileSpec tileSpec) {
```

Comment:
This is also implemented here: https://github.com/saalfeldlab/render/blob/newsolver/render-ws-java-client/src/main/java/org/janelia/render/client/multisem/Utilities.java#L61

Reply:
That's good to know, thanks! This also applies to the other PR for the re-ordering.

Reply:
The provided functionality is not the same: I really need the number of the sfov, but the given utility class also returns the scan and mfov numbers.